src/time/location.cr

```diff
@@ -0,0 +1,281 @@
+      io << ", UTC" if utc?
+      io << '>'
+    end
+  end
+
+  # Describes the Coordinated Universal Time (UTC).
+  UTC = new "UTC", [Zone::UTC]
+
+  property name : String
+  property zones : Array(Zone)
+
+  # Most lookups will be for the current time.
+  # To avoid the binary search through tx, keep a
+  # static one-element cache that gives the correct
+  # zone for the time when the Location was created.
+  # The units for @cached_range are seconds
+  # since January 1, 1970 UTC, to match the argument
+  # to `#lookup`.
+  @cached_range : Tuple(Int64, Int64)
+  @cached_zone : Zone
+
+  # Creates a `Location` instance named *name* with fixed *offset*.
+  def self.fixed(name : String, offset : Int32)
+    new name, [Zone.new(name, offset, false)]
+  end
+
+  # Creates a `Location` instance with fixed *offset*.
+  def self.fixed(offset : Int32)
+    span = offset.abs.seconds
+    name = sprintf("%s%02d:%02d", offset.sign < 0 ? '-' : '+', span.hours, span.minutes)
+    fixed name, offset
+  end
+
+  # Returns the `Location` with the given name.
+  #
+  # This uses an array of paths to look for timezone data. Each path can
+  # either point to a directory or an uncompressed ZIP file.
+  #
+  # The first timezone data matching the given name that is successfully loaded
+  # and parsed is returned.
+  #
+  # Special names:
+  # * `UTC` and empty string return `Location::UTC`
+  # * `Local` returns `Location.local`
+  #
+  # An custom lookup location can be added as environment variable `ZONEFILE`.
+  #
+  # Example:
+  # `ZONEFILE=/path/to/zoneinfo.zip crystal eval 'pp Location.load("Custom/Location")'`
+  def self.load(name : String)
+    case name
+    when "", "UTC"
+      UTC
+    when "Local"
+      local
+    when .includes?(".."), .starts_with?('/'), .starts_with?('\\')
+      # No valid IANA Time Zone name contains a single dot,
+      # much less dot dot. Likewise, none begin with a slash.
+      raise InvalidLocationNameError.new(name)
+    else
+      if zoneinfo = ENV["ZONEINFO"]?
+        if location = load_from_dir_or_zip(name, zoneinfo)
+          return location
+        else
+          raise InvalidLocationNameError.new(name, zoneinfo)
+        end
+      end
+
+      if location = load(name, Crystal::System::Time.zone_sources)
+        return location
+      end
+
+      raise InvalidLocationNameError.new(name)
+    end
+  end
+
+  # Returns the location used in the current system environment.
+  #
+  # It consults the environment variable `TZ` to find the time zone to use.
+  # * `UTC` and empty string return `UTC`
+  # * `Foo/Bar` tries to load the zoneinfo from known system locations such as `/usr/share/zoneinfo/Foo/Bar`,
+  # `/usr/share/lib/zoneinfo/Foo/Bar` or `/usr/lib/locale/TZ/Foo/Bar` on unix-based operating systems.
+  # * If `TZ` is not set, the system's local timezone data (`/etc/localtime` on unix-based systems) will be used.
+  # * If no time zone data is found, `UTC` is returned.
+  def self.local
```

Comment:
I feel like this should be cached. This method is very much on the fast path for a lot of operations.